jupyter_scheduler/task_runner.py

```diff
@@ -0,0 +1,310 @@
+                self.queue.push(
+                    JobDefinitionTask(
+                        job_definition_id=definition.job_definition_id,
+                        next_run_time=next_run_time,
+                    )
+                )
+
+    def add_job_definition(self, job_definition_id: str):
+        with self.db_session() as session:
+            definition = (
+                session.query(JobDefinition)
+                .filter(JobDefinition.job_definition_id == job_definition_id)
+                .first()
+            )
+
+        next_run_time = self.compute_next_run_time(definition.schedule, definition.timezone)
+
+        self.cache.put(
+            DescribeJobDefinitionCache(
+                job_definition_id=definition.job_definition_id,
+                active=definition.active,
+                next_run_time=next_run_time,
+                timezone=definition.timezone,
+                schedule=definition.schedule,
+            )
+        )
+        self.queue.push(
+            JobDefinitionTask(
+                job_definition_id=definition.job_definition_id, next_run_time=next_run_time
+            )
+        )
+
+    def update_job_definition(self, model: UpdateJobDefinition):
+        cache = self.cache.get(model.job_definition_id)
+        schedule = model.schedule or cache.schedule
+        timezone = model.timezone or cache.timezone
+        active = model.active if model.active is not None else cache.active
+
+        should_compute_run_time = False
+        timezone_changed = model.timezone and cache.timezone != model.timezone
+        schedule_changed = model.schedule and cache.schedule != model.schedule
+        active_changed = model.active is not None and cache.active != model.active
+
+        if active and (timezone_changed or schedule_changed):
+            should_compute_run_time = True
+
+        if active_changed and model.active:
+            should_compute_run_time = True
+
+        if should_compute_run_time:
+            next_run_time = self.compute_next_run_time(schedule, timezone)
+        else:
+            next_run_time = cache.next_run_time
+
+        if should_compute_run_time:
+            self.cache.update(
+                model.job_definition_id,
+                UpdateJobDefinitionCache(
+                    timezone=timezone, next_run_time=next_run_time, active=active, schedule=schedule
+                ),
+            )
+            self.queue.push(
+                JobDefinitionTask(
+                    job_definition_id=model.job_definition_id, next_run_time=next_run_time
+                )
+            )
+        elif active_changed and not active:
+            self.cache.update(model.job_definition_id, UpdateJobDefinitionCache(active=active))
+        elif schedule_changed or timezone_changed:
+            self.cache.update(
+                model.job_definition_id,
+                UpdateJobDefinitionCache(timezone=timezone, schedule=schedule),
+            )
```

Comment:
I'm not sure why this method is so overly verbose. I suggest we rewrite this. Here's the flow the update path should take:

1. When we update a job definition, we **always** want to update the cache (i.e. it is a write-through cache). We do not need or want to try to be intelligent about when we're doing so, since that opens us up to possible cache inconsistency. For example, what if a new field was needed by the task runner and needed to be added to the job definition cache? If you forgot to update this specific method, you would have cache inconsistency, because you're only updating the cache if any of the previous cache fields changed.

2. To determine whether we add this job definition into the queue, we just need to check if the previous `next_run_time` and new `next_run_time` differ, and if the new job definition is active.

Rough draft for what you would want here:

```py
cache = self.cache.get(model.job_definition_id)
next_run_time = self.compute_next_run_time(model.schedule, model.timezone)

# always update write-through cache
self.cache.update(model.job_definition_id, UpdateJobDefinitionCache(**model))

# if next run time differs, and new job definition is active, add it to the queue
if cache.next_run_time != next_run_time and model.active:
  self.queue.push(JobDefinitionTask(job_definition_id=model.job_definition_id, next_run_time=next_run_time))
```

Reply:
This is a good point, but something to note here is that `model` here might not have any of the 3 values we are interested in here. Also, relying on the `next_run_time` for comparison means we might be pushing the next set of tasks to the queue even if  the `schedule` or `timezone` hasn't changed; I am not sure this comparison will work for all cron expressions.

Reply:
Oh right, the update model is a partial of the job definition model. My bad, you do actually want to include the below logic in your implementation:

```
        schedule = model.schedule or cache.schedule
        timezone = model.timezone or cache.timezone
```

However, there is another issue here that's inherent to how we are handling update API requests. Because `model.schedule` allows `None` for a value, it seems impossible to distinguish from the `UpdateJobDefinition` model whether the user left the field unspecified, or if they specified the field to be `None` (e.g. the difference in JSON request bodies is `{}` v.s. `{ schedule: null }`). In both cases, `model.schedule` is set to `None`.

> Also, relying on the next_run_time for comparison means we might be pushing the next set of tasks to the queue even if the schedule or timezone hasn't changed; I am not sure this comparison will work for all cron expressions.

This should never happen, since if `schedule` and `timezone` have not changed, then `next_run_time` should not change either.

Reply:
The new logic looks great. 👍

Reply:
Tracking in #119.